device_doctor/lib/src/android_device.dart

```diff
@@ -0,0 +1,195 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+import 'package:retry/retry.dart';
+
+import 'device.dart';
+import 'process_helper.dart';
+import 'health.dart';
+import 'utils.dart';
+
+/// Constant battery health values returned from Android Battery Manager.
+///
+/// https://developer.android.com/reference/android/os/BatteryManager.html
+class AndroidBatteryHealth {
+  // Match SCREAMING_CAPS to Android constants.
+  static const BATTERY_HEALTH_UNKNOWN = 1;
+  static const BATTERY_HEALTH_GOOD = 2;
+  static const BATTERY_HEALTH_OVERHEAT = 3;
+  static const BATTERY_HEALTH_DEAD = 4;
+  static const BATTERY_HEALTH_OVER_VOLTAGE = 5;
+  static const BATTERY_HEALTH_UNSPECIFIED_FAILURE = 6;
+  static const BATTERY_HEALTH_COLD = 7;
+}
+
+class AndroidDeviceDiscovery implements DeviceDiscovery {
+  factory AndroidDeviceDiscovery() {
+    return _instance ??= AndroidDeviceDiscovery._();
+  }
+  AndroidDeviceDiscovery._();
+
+  @visibleForTesting
+  AndroidDeviceDiscovery.testing();
+
+  // Parses information about a device. Example:
+  //
+  // 015d172c98400a03       device usb:340787200X product:nakasi model:Nexus_7 device:grouper
+  static final RegExp _kDeviceRegex = RegExp(r'^(\S+)\s+(\S+)(.*)');
+
+  static AndroidDeviceDiscovery _instance;
+
+  Future<String> deviceListOutput() async {
+    return eval(properties['adb'], <String>['devices', '-l'], canFail: false).timeout(Duration(seconds: 15));
+  }
+
+  Future<List<String>> deviceListOutputWithRetries(Duration retriesDelayMs) async {
+    RetryOptions r = RetryOptions(
+      maxAttempts: 3,
+      delayFactor: retriesDelayMs,
+    );
+    return await r.retry(
+      () async {
+        String result = await deviceListOutput();
+        return result.trim().split('\n');
+      },
+      retryIf: (Exception e) => e is TimeoutException,
+      //onRetry: (Exception e) => killAdbServer(),
+    );
+  }
+
+  void killAdbServer() async {
```

Comment:
nit: ADB related calls are implementation details to this class and should be hidden

Reply:
This needs to be implemented in fake_android_device for testing purpose.

Reply:
In that case, it sounds like you want a dedicated Adb class that you can create a fake for

Reply:
Have refactored class to test the whole logic by injecting progress manager, hiding related functionsl